torch/csrc/jit/custom_graph_executor_impl.cpp

```diff
@@ -0,0 +1,60 @@
+#include <torch/csrc/jit/custom_graph_executor_impl.h>
+
+namespace torch {
+namespace jit {
+
+extern const Symbol kDefaultExecutor =
+    Symbol::fromQualString("executor::default");
+
+namespace {
+
+struct GraphExecutorImplState {
+  Symbol name = kDefaultExecutor;
+  bool impl_cached = false;
+  GraphExecutorImplCreator impl_fn;
+};
+thread_local GraphExecutorImplState executor_state;
+
+/* Meyers Singleton */
+std::unordered_map<Symbol, GraphExecutorImplCreator>& getGraphExecutorImpls() {
+  static std::unordered_map<Symbol, GraphExecutorImplCreator> executors;
+  return executors;
+}
+
+} // namespace
+
+Symbol getGraphExecutorName() {
+  return executor_state.name;
+}
+
+void setGraphExecutorName(Symbol name) {
+  auto& executors = getGraphExecutorImpls();
+  auto it = executors.find(name);
+  TORCH_CHECK(
+      it != executors.end(),
+      "No GraphExecutorImpl registered for ",
+      name.toDisplayString());
+
+  executor_state.name = name;
+  executor_state.impl_cached = true;
+  executor_state.impl_fn = it->second;
+}
+
+RegisterGraphExecutorImpl::RegisterGraphExecutorImpl(
+    Symbol name,
+    GraphExecutorImplCreator creator) {
+  auto& executors = getGraphExecutorImpls();
+  TORCH_CHECK(
+      executors.emplace(name, creator).second,
+      "Cannot register GraphExecutorImpl for ",
```

Comment:
A better error message would be something along the lines "GraphExecutorImpl for ... has already been registered!"